comdb2/_ccdb2.pyx

```diff
@@ -153,6 +160,54 @@ cdef class _ParameterValue(object):
                 self.data = PyMem_Malloc(self.size)
                 _bind_datetime(obj, <lib.cdb2_client_datetime_t*>self.data)
                 return
+            elif isinstance(obj, (list, tuple)):
+                self.list_size = len(obj)
+                if 0 == self.list_size:
+                    raise ValueError(f"empty {type(obj).__name__}s cannot be bound")
+
+                if all(isinstance(ele, six.integer_types) for ele in obj):
+                    self.type = lib.CDB2_INTEGER
+                    self.size = sizeof(long long)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
+                    for l_index in range(self.list_size):
+                        (<long long*>self.data)[l_index] = obj[l_index]
+                    return
+                elif all(isinstance(ele, float) for ele in obj):
+                    self.type = lib.CDB2_REAL
+                    self.size = sizeof(double)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
+                    for l_index in range(self.list_size):
+                        (<double*>self.data)[l_index] = obj[l_index]
+                    return
+                elif all(isinstance(ele, bytes) for ele in obj):
+                    self.type = lib.CDB2_BLOB
+                    self.size = sizeof(blob_descriptor)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
```

Comment:
Why call `PyMem_Malloc` here? Aren't we using twice as much memory as we need to?

I think `self.owner` could be initialized to `obj`, which would increase the reference count by one and ensure it doesn't get garbage collected until nothing has a reference to the `_ParameterValue` instance, and we can just make `self.data` point to the data held by `self.owner`.

Reply:
> Why call `PyMem_Malloc` here? Aren't we using twice as much memory as we need to?

We need to pass a contiguous array of `size_t` / `char*` pairs to libcdb2api, and that's not something that we can export from the CPython C list/tuple API - we need to allocate memory for that array, and free it when we're done.

> I think `self.owner` could be initialized to `obj`, which would increase the reference count by one and ensure it doesn't get garbage collected until nothing has a reference to the `_ParameterValue` instance

I was going to say that we don't need to worry about this - that `parameters` itself outlives the `_ParameterValue` instances, and would keep `obj` alive. But, that's not quite true: while we're making the call to `cdb2_run_statement` we drop the GIL, and once we drop the GIL another thread could mutate the `parameters` dict so that it no longer references `obj`. So, I think you're half right - we _do_ need to set `self.owner = obj` here, even though we _also_ need to malloc a contiguous array.

Reply:
Yeah, I guess it got overridden during Matt's changes. Adding it back.